### DISCUSSION TITLE
Using sniprun to render plantuml diagrams

### DISCUSSION THREAD

**Discussion by Piotr1215** [michaelb/sniprun#289] [Q&A] [answered]
EDIT: `spinrun` -> `sniprun`

I would like to use `sniprun` to render a `PlantUML` diagram and link the resulting *.png file or ascii right below the code fence block with the diagram content. 

Has anyone figured out if it's possible and how to do it?

**Comment by michaelb** [comment, marked as answer]
It's 'sniprun' haha

outputting a png is definitely not supported, and to even embed it in your terminal would require support and special integration with that terminal application.

ASCII is in the realm of the possible. Do you know of any (command-line) application that converts a file containing a valid plantulm diagram into some ascii representation ?

If so, the [Generic interpreter](https://michaelb.github.io/sniprun/sources/interpreters/Generic.html) can most likely be configured to use that.

In addition, if you don't like sniprun's display modes (which may not be a perfect fit for displaying ascii figures), you may have to write your own output code using the [API display mode](https://michaelb.github.io/sniprun/sources/display_modes.html#api-display)

Good luck

**Reply by Piotr1215** [reply]
Och lol, I was working on a `spin operator` when writing haha.

I have this autocommand that generates *.svg on *.puml save

```lua
api.nvim_exec(
  [[
    augroup plantuml
     autocmd BufWritePost *.puml silent! !java -DPLANTUML_LIMIT_SIZE=8192 -jar /usr/local/bin/plantuml.jar -tsvg <afile> -o ./rendered
    augroup end
  ]],
  false
)
```

It can be tweaked to output `-tutxt` for example. The extensibility model should help here, need to queue a todo for it.

**Reply by michaelb** [reply]
I actually took the time to write a plantuml interpreter. It's currently on the dev branch, if you're willing to take it for a spin. I can't promise it's perfect just yet though.

**Reply by Piotr1215** [reply]
Thank you so much for taking the time to put this together. I've installed the plugin from dev branch with 
`use { "michaelb/sniprun", run = "sh ./install.sh", branch = "dev" }` and checked that dev branch was pulled.

Tested on a few examples form my [diagrams repo](https://github.com/Piotr1215/cloudrumble/tree/master/diagrams), but keep getting:
 `Filetype not officially supported, nor configured for the Generic interpreter`.

`SnipInfo` output doesn't include plantuml.

I have included the settings for Plantuml in the plugin load:

```lua
require("sniprun").setup {
  interpreter_options = {
    Plantuml_original = {
      output_mode = "-tutxt", --# one of the options allowed by plantuml
    },
  },
}
```

Just to clarify, the tests were:
1. open `*.puml` file, run `SnipRun` or select all and run `SnipRun` on the selection
2. create code fenced snipped with the diagram code in a `markdown` file and to operations from point 1.

**Reply by michaelb** [reply]
Okay, the missing item in snipinfo is not supposed to happen ....strange

What's the actual filetype, for reference? (You can get it by executing `:set ft?` in neovim

I think sniprun didn't properly update: you may have switched branches but I forgot to tell you you'll need a Rust toolchain so sniprun hets compiled locally. You're probably using the old binary compiled from the master branch (actually, fetched from the github releases)

If you don't want to install a rust toolchain & cie, no worries, but you'll have to wait for the next release

**Reply by Piotr1215** [reply]
I have rust tool chain installed. 

```bash
rustup 1.26.0 (5af9b9484 2023-04-05)
info: This is the version for the rustup toolchain manager, not the rustc compiler.
info: The currently active `rustc` version is `rustc 1.75.0 (82e1608df 2023-12-21)`
```

When running `./install.sh` locally, I'm getting the following error, so you might be right, it fetches existing binary from the GitHub releases. 

```bash
➜ ./install.sh
Runnning Sniprun Installer
./install.sh: 8: [: dev: unexpected operator
Downloading sniprun binary: v1.3.14
Done
```

Running the installer via bash `bash +x ./install.sh` triggers local build with `cargo` and I was able to run the dev version.

I believe the issue is that `sh` interpreter doesn't work with double  `==` for comparison operator, this:

```bash
current_branch=$(git rev-parse --abbrev-ref HEAD)
if [ "$current_branch" = "dev" ]; then
	force_build=1
fi
```
fixes the issue. Created a PR: https://github.com/michaelb/sniprun/pull/291

Some diagrams work, some produce java and other errors. The errors with diagrams 2 and 3 might be due to the output, generally `-utxt` is supported only for sequence diagrams.

1. [Works](https://github.com/Piotr1215/cloudrumble/blob/master/diagrams/git-workflow.puml), however the output is in the temporary window, I was expecting it to be directly under the diagram.
2. [Java Error](https://github.com/Piotr1215/cloudrumble/blob/master/diagrams/k8s-architecture.puml)
```bash
java.lang.IllegalArgumentException
        at net.sourceforge.plantuml.posimo.Path.<init>(Path.java:52)
        at net.sourceforge.plantuml.cucadiagram.dot.CucaDiagramTxtMaker.<init>(CucaDiagramTxtMaker.java:113)
        at net.sourceforge.plantuml.cucadiagram.CucaDiagram.createFilesTxt(CucaDiagram.java:675)
        at net.sourceforge.plantuml.cucadiagram.CucaDiagram.exportDiagramInternal(CucaDiagram.java:625)
        at net.sourceforge.plantuml.UmlDiagram.exportDiagramNow(UmlDiagram.java:201)
        at net.sourceforge.plantuml.AbstractPSystem.exportDiagram(AbstractPSystem.java:135)
        at net.sourceforge.plantuml.PSystemUtils.exportDiagramsCuca(PSystemUtils.java:235)
        at net.sourceforge.plantuml.PSystemUtils.exportDiagrams(PSystemUtils.java:91)
        at net.sourceforge.plantuml.SourceFileReaderAbstract.getGeneratedImages(SourceFileReaderAbstract.java:159)
        at net.sourceforge.plantuml.Run.manageFileInternal(Run.java:507)
        at net.sourceforge.plantuml.Run.access$000(Run.java:77)
        at net.sourceforge.plantuml.Run$1.run(Run.java:418)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
```
3. C4 diagrams don't work, like [this one](https://github.com/Piotr1215/cloudrumble/blob/master/diagrams/vcluster-components.puml)

**Reply by michaelb** [reply]
> When running ./install.sh locally, I'm getting the following error, so you might be right, it fetches existing binary from the GitHub releases.

Thank you for this feedback (it's quite precious as I can't have all the bash/sh/zsh versions in the world on my one PC)

> 1. 

You can configure the output (for example to a [floating window](https://michaelb.github.io/sniprun/sources/display_modes.html#temporary-floating-windows). However, with the sizes of those diagrams, we're gonna have a problem anyway. People have experimented writing results in the buffer, but this is not officially supported: https://github.com/michaelb/sniprun/discussions/128.

If there is a way to display the diagrams somehow in a graphical window, alike a python matplotlib.plot, it should be possible to have that as output by adding the relevant plantuml options in sniprun's configuration ('compiler' key; see :SnipInfo Plantuml_original)

> 2. 

I have the same issue inside sniprun and on the CLI (actually, a NullPointerException, maybe due to different plantuml version - 1.2023.13 - or arguments - see `:SnipInfo Plantuml_original` - ?) Maybe the diagram is somehow invalid ?

> 3.

Surprisingly, it works on my machine.. Maybe again a different plantuml version/config is the cause ? It's a bit slow, as any good plantuml diagram ought to be, but nothing extraordinary either.

**Reply by Piotr1215** [reply]
No worries, happy I could help. I also realized I created a PR to main and not dev ;)

My use case is mostly for `markdown`. I would like to have a `\```plantuml` code fence with the diagram and right below it render it as an actual ASCII diagram (again this would work only with sequence diagrams). Similarly like [mdeval.nvim](https://github.com/jubnzv/mdeval.nvim) works. Would something like this be possible with `sniprun`?

**Reply by michaelb** [reply]
Yes, it's possible ! Actually, the markdown (sniprun) interpreter already inherits support and features of all languages, so it'll work out of the box, similarly to mdeval.nvim *!

![Screenshot_%T_%d-19](https://github.com/user-attachments/assets/cc5d107a-2294-4f66-b6ab-9af0b6351f79)


*Aside the display of the result, which you'll have to choose according to your preference and the constraints of the (sometime huge) diagrams you'll want to work with. The screenshot is using 'TempFloatingWindow'

**Reply by Piotr1215** [reply]
Great, thank you! This will be very helpful for markdown based terminal slides I use very often. Looking forward to the new release with `PlantUML` support 🤝

**Reply by michaelb** [reply]
v1.3.15 is out with plantUML support :-)

**Reply by Piotr1215** [reply]
Thank you! It works without any additional config. There is a small glitch, after executing `:SpinRun` a test PlantUML markdown snippet:

\```plantuml
@startuml
Alice -> Bob: Authentication Request
Bob --> Alice: Authentication Response

Alice -> Bob: Another authentication Request
Alice <-- Bob: Another authentication Response
@enduml
\```

getting this: 
![image](https://github.com/user-attachments/assets/d4c0d46d-2926-45ed-a38f-472871ca5ddf)

It disappears on file reload or `:SnipClose`

**Reply by michaelb** [reply]
> Thank you! It works without any additional config. There is a small glitch, after executing `:SpinRun` a test PlantUML markdown snippet:
> 
> \```plantuml
> @startuml
> Alice -> Bob: Authentication Request
> Bob --> Alice: Authentication Response
> 
> Alice -> Bob: Another authentication Request
> Alice <-- Bob: Another authentication Response
> @enduml
> \```
> 
> getting this: 
> ![image](https://github.com/user-attachments/assets/d4c0d46d-2926-45ed-a38f-472871ca5ddf)
> 
> It disappears on file reload or `:SnipClose`
> 

Nice :-)
Oh haha it's not a glitch, it's the VirtualText display mode: sniprun defaults to displaying with the 'Classic' (in the command area) and the 'VirtualTextOk' (successful results display over this 1-line space). For plantuml however, this is quite ill-suited. Defining the 'display' key in your configuration will overwrite the default with whatever mode you fancy 
https://michaelb.github.io/sniprun/sources/display_modes.html#available-display-modes

Also, if you want the whole snippet to be sent and not just Alice->Bob, you can run with the cursor either on \@startuml or the markdown code bloc marker

**Reply by Piotr1215** [reply]
Ah, that makes sense. Both the `terminal` and `classic` settings work here pretty well. 

Thank you for implementing the PlantUML runner and for the great plugin, I'm sure I've just scratched the surface with my use case.

A bit off-topic, but it's great to see a remote plugin mixing `rust` and `lua`. I've been playing with the idea of creating a kind of neovim slides plugin similar to `lookatme` or `slides`, but fully encapsulated in neovim. Looking at you repo, rust might be a neat idea for the project!

**Reply by michaelb** [reply]
> Ah, that makes sense. Both the `terminal` and `classic` settings work here pretty well. 
> 
> Thank you for implementing the PlantUML runner and for the great plugin, I'm sure I've just scratched the surface with my use case.
> 
> A bit off-topic, but it's great to see a remote plugin mixing `rust` and `lua`. I've been playing with the idea of creating a kind of neovim slides plugin similar to `lookatme` or `slides`, but fully encapsulated in neovim. Looking at you repo, rust might be a neat idea for the project!

I'd lie if I'd say it's like walking on a red carpet. The tooling has evolved somewhat since I created sniprun (we now have a supposedly nicer async rpc framework 'nvim-rs') but overall the examples/docs are scarse.

Rust(+lua) luckily struck a good balance of complexity (of the tools) versus the needs of sniprun itself (in terms of overall structure, maintainability, expendability, reliability etc..)
At the time lua was barely a thing (🦕), and VimScript clearly wasn't gonna cut it past a couple hundred lines of code. While if I had to redo it now, I still would choose to do so in Rust, for a plugin smaller in scope 'pure' Lua would probably be my go-to
